operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -607,133 +198,125 @@ func (c *cesManagerIdentity) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint,
-// api-server to cache. In this case, isDeepCopy set to true to copy entire CEP object locally.
-// 2) During runtime, reconciler sync current state with API server and update metadata only.
-// isDeepCopy flag is set to false.
-func (c *cesManagerIdentity) updateCESInCache(srcCES *cilium_v2.CiliumEndpointSlice, isDeepCopy bool) {
-	if ces, ok := c.desiredCESs.getCESTracker(srcCES.GetName()); ok {
-		// The identityLock and backendMutex locks always need to be acquired in the
-		// same order to avoid deadlocks. First identityLock and then backendMutex,
-		// because identityLock is a higher level lock of cesManagerIdentity which
-		// contains cesTrackers with backendMutex locks within it.
-		c.identityLock.Lock()
-		defer c.identityLock.Unlock()
-		ces.backendMutex.Lock()
-		defer ces.backendMutex.Unlock()
-		if !isDeepCopy {
-			ces.ces.ObjectMeta = srcCES.ObjectMeta
-		} else {
-			ces.ces = srcCES
-			_, exist := c.cesToIdentity[srcCES.GetName()]
-			for _, cep := range ces.ces.Endpoints {
-				// Update the identityToCES and cesToIdentity maps respectively.
-				if !exist {
-					c.identityToCES[cep.IdentityID] = append(c.identityToCES[cep.IdentityID], ces)
-					c.cesToIdentity[srcCES.GetName()] = cep.IdentityID
-					exist = true
-				}
-				// Update the desiredCESs, to reflect all CEPs are packed in a CES
-				c.desiredCESs.insertCEP(GetCEPNameFromCCEP(&cep, ces.ces.Namespace), srcCES.GetName())
+func (c *cesManagerIdentity) RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns string) CESName {
+	cepName := GetCEPNameFromCCEP(cep, ns)
+	c.logger.WithFields(logrus.Fields{
+		logfields.CEPName: cepName.string(),
+	}).Debug("Removing CEP from local cache")
+	cesName, exists := c.mapping.getCESName(cepName)
+	if exists {
+		c.logger.WithFields(logrus.Fields{
+			logfields.CEPName: cepName.string(),
+			logfields.CESName: cesName.string(),
+		}).Debug("Removing CEP from CES")
+		c.mapping.deleteCEP(cepName)
+		if c.mapping.countCEPsInCES(cesName) == 0 {
+			c.mutex.Lock()
+			defer c.mutex.Unlock()
+			if c.mapping.countCEPsInCES(cesName) == 0 {
+				c.removeCESToIdentity(cep.IdentityID, cesName)
+				c.mapping.deleteCES(cesName)
 			}
 		}
-	} else {
-		log.WithFields(logrus.Fields{
-			logfields.CESName: srcCES.GetName(),
-		}).Debug("Attempted to updateCESInCache non-existent cesName , skipping")
+		return cesName
 	}
+	return CESName(resource.Key{})
 }
 
-// Insert the ces in workqueue
-func (c *cesMgr) insertCESInWorkQueue(ces *cesTracker, baseDelay time.Duration) {
-	// If CES insert time is not zero, save current time.
-	if ces.cesInsertedAt.IsZero() {
-		ces.cesInsertedAt = time.Now()
+func (c *cesManagerIdentity) removeCESToIdentity(id int64, cesName CESName) {
+	cesSlice := c.identityToCES[id]
+	for i, ces := range cesSlice {
+		if ces == cesName {
+			cesSlice[i] = cesSlice[len(cesSlice)-1]
+		}
 	}
+	c.identityToCES[id] = cesSlice
```

Comment:
Shouldn't we keep track of how many elements we've overwritten in cesSlice at line 291 and then assign here `cesSlice[:len(cesSlice)-removed]? Otherwise the slices in the `c.identityToCES` won't never shrink.

Reply:
Fixed. Good catch.